csharp/src/Microsoft.ML.OnnxRuntime/OrtAllocator.cs

```diff
@@ -0,0 +1,406 @@
+            return allocatorType;
+        }
+
+        public bool CompareMemoryInfo(OrtMemoryInfo other)
+        {
+            int result = -1;
+            NativeApiStatus.VerifySuccess(NativeMethods.OrtCompareMemoryInfo(_pointer, other._pointer, out result));
+            return (result == 0);
+        }
+        #region IDisposable Support
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                if (_owned)
+                {
+                    NativeMethods.OrtReleaseMemoryInfo(_pointer);
+                }
+                _pointer = IntPtr.Zero;
+            }
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+        // We intentionally do not provider an finalizer for the class
+        #endregion
+    }
+
+    /// <summary>
+    /// This class represents memory allocation made by a specific onnxruntime
+    /// allocator. Use OrtAllocator.Allocate() to obtain an instance of this class.
+    /// It implements IDisposable and makes use of the original allocator
+    /// used to allocate the memory. The lifespan of the allocator instance must eclipse the
+    /// lifespan of the allocation. Or, if you prefer, all OrtMemoryAllocation instances must be
+    /// disposed of before the corresponding allocator instances are disposed of.
+    /// </summary>
+    public class OrtMemoryAllocation : IDisposable
+    {
+        private OrtAllocator _allocator;
+
+        /// <summary>
+        /// Bind an arbitrary piece of native memory to the instance
+        /// The instance will not have the ownership of this memory.
+        /// </summary>
+        /// <param name="pointer"></param>
+        /// <param name="size"></param>
+        public OrtMemoryAllocation(IntPtr pointer, uint size)
+        {
+            _allocator = null;
+            Pointer = pointer;
+            Size = size;
+        }
+
+        /// <summary>
+        /// This constructs an instance representing an native memory allocation.
+        /// Typically returned by OrtAllocator.Allocate(). However, some APIs return
+        /// natively allocated IntPtr using a specific allocator. It is a good practice
+        /// to wrap such a memory into OrtAllocation for proper disposal. You can set
+        /// size to zero if not known, which is not important for disposing.
+        /// </summary>
+        /// <param name="allocator"></param>
+        /// <param name="pointer"></param>
+        /// <param name="size"></param>
+        public OrtMemoryAllocation(OrtAllocator allocator, IntPtr pointer, uint size)
+        {
+            _allocator = allocator;
+            Pointer = pointer;
+            Size = size;
+        }
+
+        /// <summary>
+        /// Internal accessor to call native methods
+        /// </summary>
+        internal IntPtr Pointer { get; private set; }
```

Comment:
This is internal, so it would be not accessible to the end users. We would like to use this the same way you use it in InferenceTest.cs::PopulateNativeBufferFloat(). You could provide a guarded read/write functions here, which would check for the size.

Reply:
This is the topic that is under discussion, earlier you mentioned that your current use case is satisfied w/o reading/writing native allocations. If not, let's discuss it and we will provide a way. There is not a way back when a native ptr is exposed.

---
In reply to: [463175581](https://github.com/microsoft/onnxruntime/pull/4646#discussion_r463175581) [](ancestors = 463175581)